src/utils/wallets/wallets.utils.ts

```diff
@@ -0,0 +1,409 @@
+import * as bip39 from "bip39-web-crypto";
+import {
+  addWallet,
+  getWalletKeyLength,
+  updatePassword,
+  type WalletKeyLengths
+} from "~wallets";
+import {
+  checkPasswordValid,
+  jwkFromMnemonic,
+  pkcs8ToJwk
+} from "~wallets/generator";
+import * as SSS from "shamir-secret-sharing";
+import type { JWKInterface } from "arweave/web/lib/wallet";
+import { defaultGateway } from "~gateways/gateway";
+import Arweave from "arweave";
+import { nanoid } from "nanoid";
+import { setDecryptionKey } from "~wallets/auth";
+import {
+  INVALID_DEVICE_NONCE_ERR_MSG,
+  INVALID_DEVICE_SHARES_INFO_ERR_MSG
+} from "~utils/wallets/wallets.constants";
+import { ExtensionStorage } from "~utils/storage";
+import { log, LOG_GROUP } from "~utils/log/log.utils";
+
+async function generateSeedPhrase() {
+  log(LOG_GROUP.WALLET_GENERATION, "generateSeedPhrase()");
+
+  return bip39.generateMnemonic();
+}
+
+async function generateWalletJWK(seedPhrase: string): Promise<JWKInterface> {
+  if (!seedPhrase) throw new Error("Missing `seedPhrase`");
+
+  log(LOG_GROUP.WALLET_GENERATION, "generateWalletJWK()");
+
+  let generatedKeyfile: JWKInterface | null = null;
+  let walletKeyLength: WalletKeyLengths | null = null;
+
+  let attempts = 0;
+
+  // This do-while is used to just to make sure the key has the right length, as we had some reports
+  // in the past of people having RSA-2048. In any case, this should never run more than once.
+  do {
+    ++attempts;
+
+    generatedKeyfile = await jwkFromMnemonic(seedPhrase);
+
+    walletKeyLength = await getWalletKeyLength(generatedKeyfile);
+  } while (!generatedKeyfile || !walletKeyLength.match);
+
+  if (attempts > 1) {
+    // TODO: Send this to Sentry or whatever...
+    console.warn(
+      "Took multiple attempts to generate a wallet with the right length!"
+    );
+  }
+
+  return generatedKeyfile;
+}
+
+export interface WorkShares {
+  authShare: string;
+  deviceShare: string;
+}
+
+async function generateWalletWorkShares(
+  jwk: JWKInterface
+): Promise<WorkShares> {
+  log(LOG_GROUP.WALLET_GENERATION, "generateWalletWorkShares()");
+
+  const privateKeyJWK = await window.crypto.subtle.importKey(
+    "jwk",
+    jwk,
+    { name: "RSA-PSS", hash: "SHA-256" },
+    true,
+    ["sign"]
+  );
+
+  const privateKeyPKCS8 = await window.crypto.subtle.exportKey(
+    "pkcs8",
+    privateKeyJWK
+  );
+
+  // Wanna know why these are called "shares" and not shards?
+  // See https://discuss.hashicorp.com/t/is-it-shards-or-shares-in-shamir-secret-sharing/38978/3
+
+  const [authShareBuffer, deviceShareBuffer] = await SSS.split(
+    new Uint8Array(privateKeyPKCS8),
+    2,
+    2
+  );
+
+  // TODO: Need to add Buffer polyfill
```

Comment:
Isn't this already addressed with nodePolyFills?

Reply:
True, I added this TODO really early on and forgot about it. Let me remove it.